pywhy_graphs/algorithms/pag.py

```diff
@@ -1176,3 +1183,206 @@ def pag_to_mag(graph):
+
+    # temp_pag.add_edges_from(undedges, temp_pag.undirected_edge_name)
+
+    temp_pag.add_edges_from(biedges, temp_pag.bidirected_edge_name)
+
+    all_nodes = set(temp_pag.nodes)
+
+    for source in all_nodes:
+        nb = set(temp_pag.neighbors(source))
+        cur_set = all_nodes - nb
+        cur_set.remove(source)
+        for dest in cur_set:
+            out = inducing_path(temp_pag, source, dest, L, S)
+            if out[0] is True:
+                return False
+
+    return True
+
+
+def mag_to_pag(G: PAG):
+    """Converts the provided mag into a pag using the FCI algorithm.
+
+    The FCI algorithms, as defined in :footcite:`Zhang2008` is a provably
+    complete for learning all the tractable features of an MAG, thus
+    producing a PAG.
+
+    Parameters
+    ----------
+    G : MAG
+        The MAG.
+
+    Returns
+    -------
+    pag : PAG
+        The PAG constructed from the MAG.
+    """
+    try:
+        from dodiscover import FCI, make_context
+        from dodiscover.ci import Oracle
+        from dodiscover.constraint.utils import dummy_sample
+    except ImportError as e:
+        raise ImportError("The 'dodiscover' package is required to convert a MAG to a PAG.")
+
+    data = dummy_sample(G)
+    oracle = Oracle(G)
+    # ci_estimator = GSquareCITest(data_type="discrete")
+    context = make_context().variables(data=data).build()
+    fci = FCI(ci_estimator=oracle)
+    fci.learn_graph(data, context)
+
+    return fci.graph_
+
+
+def _check_pag_edges_are_equal(G1: PAG, G2: PAG):
+    """Check if the two provided PAGs are equivalent or not.
+
+    This function compares the edges in both the graphs to determine
+    equivalency.
+
+    Parameters
+    ----------
+    G1 : PAG
+        The first PAG.
+
+    G2 : PAG
+        The second PAG.
+
+    Returns
+    -------
+    is_equivalent : bool
+        A boolean indicating whether the two PAGs are equivalent or not.
+    """
+
+    g1_edges = G1.edges()
+    g2_edges = G2.edges()
+
+    if set(g1_edges["directed"]) != set(g2_edges["directed"]):
+        return False
+
+    elif set(g1_edges["undirected"]) != set(g2_edges["undirected"]):
+        return False
+
+    elif set(g1_edges["bidirected"]) != set(g2_edges["bidirected"]):
+        return False
+
+    elif set(g1_edges["circle"]) != set(g2_edges["circle"]):
+        return False
+
+    else:
+        return True
+
+
+def valid_pag(G: PAG):
+    """Check if the provided PAG is valid or not.
+
+    The function determines the validity by first converting the PAG
+    into an MAG, then checking the validity of the said MAG. After the
+    validity of the MAG has been established, the MAG is converted back
+    into a PAG. Then the function checks to see if the original and the
+    reconverted PAG are equivalent or not.
+
+    Parameters
+    ----------
+    G : PAG
+        The PAG.
+
+    Returns
+    -------
+    is_valid : bool
+        Boolean indicating whether the provided PAG is valid or not.
```

Comment:
Can you add a `References` section, so the bibliography for Zhang2008 renders here?

Reply:
Alright.